### ISSUE TITLE
Allow observer nodes to join a running network.

### ISSUE THREAD

**Issue by afck** [poanetwork/hbbft#94]
For completely new nodes, joining a running consensus session will involve two steps:
* Join as an observer, so that there is a clearly defined first epoch from which they are guaranteed to receive all output.
* Then perform [Distributed Key Generation](https://github.com/poanetwork/hbbft/issues/47) and turn into a full peer that takes part in the consensus itself.

With #93, it will be possible to remove peers and turn them into observer nodes, and to turn existing observer nodes into peers, so the second point is basically done.

For the first point, the signal indicating the new observer has to be defined by transactions itself — it already is, in #93, namely a majority of `Add` votes, which is communicated to the user as part of the `Batch`. The user is responsible for making sure that the new node receives all messages after that output. Once #91 is done, that guarantees that the new node indeed receives all messages for later epochs.

Finally, we need to modify the `HoneyBadger` and `DynamicHoneyBadger` constructors to allow starting in an epoch other than `0`, and write more tests for changing peer/validator sets.

**Comment by afck** [comment]
This is basically done, with the `JoinPlan`.
It would be nice to allow nodes to join in any epoch and not only after the ones with a change entry, but that's probably a separate issue.
